ecl/hql/hqlcache.cpp

```diff
@@ -340,6 +342,141 @@ void convertSelectsToPath(StringBuffer & filename, const char * eclPath)
+                                }
+                                else
+                                {
+                                    IValue * val = attr->queryValue();
+                                    if (!val)
+                                        throw MakeStringExceptionDirect(0, "Cannot simplify record");
+                                    //For numeric constants values of attributes, it is not possible to specify the type of the number
+                                    //in the generated cache.  The parser treats numbers as INTEGER8 by default.  Here, regardless of
+                                    //the exact numeric type of an attribute's value, generate it as the default integer type (INTEGER8)
+                                    if (val->getTypeCode()==type_int)
+                                        attrs.append(*createConstant(val->getIntValue()));
+                                    else
+                                        attrs.append(*createConstant(LINK(attr)));
+                                    same2 = false;
+                                }
+                            }
+                            if (same2)
+                            {
+                                children.append(*(LINK(cur)));
+                            }
+                            else
+                            {
+                                OwnedHqlExpr newattr = cur->queryBody()->clone(attrs);
+                                children.append(*(newattr.getClear()));
+                                same = false;
+                            }
+                        }
+                    }
+                    else
+                        same = false;
+                }
+                OwnedITypeInfo newType = transformType(type);
+
+                if (type != newType)
+                    return createField(expr->queryId(), newType.getClear(), children);
+
+                if (same)
+                    return LINK(expr);
+                return expr->clone(children);
+            }
+        case no_param:
+            {
+                ITypeInfo * type = expr->queryType();
+                OwnedITypeInfo newType = transformType(type);
+                HqlExprArray children;
+                transformChildren(expr, children);      // could just unwind
+                return createParameter(expr->queryId(), UnadornedParameterIndex, newType.getClear(), children);
+            }
+        }
+
+        return QuickHqlTransformer::createTransformed(expr);
+    }
+
+protected:
+    IHqlExpression * createTransformedField(IHqlExpression * expr)
+    {
+        Owned<ITypeInfo> type = getFullyUnqualifiedType(expr->queryType());
+        OwnedITypeInfo newType = transformType(type);
+        IIdAtom * id = expr->queryId();
+        return createField(id, newType.getClear(), nullptr);
+    }
+    ITypeInfo * transformType(ITypeInfo * type)
```

Comment:
should be marked as virtual, and these days we should really use override on all the overridden functions (this and above).